components/net/fetch/methods.rs

```diff
@@ -732,9 +735,30 @@ async fn scheme_fetch(
                 if let Ok(file) = File::open(file_path.clone()) {
                     if let Ok(metadata) = file.metadata() {
                         if metadata.is_dir() {
-                            return Response::network_error(NetworkError::Internal(
-                                "Opening a directory is not supported".into(),
-                            ));
+                            // Skip past the "file://" prefix, otherwise from_directory_path
+                            // will return an error.
+                            let url_path = &url.as_str()[7..];
+                            let url = if !url_path.ends_with('/') {
+                                if let Ok(dir_url) = Url::from_directory_path(url_path) {
+                                    ServoUrl::from_url(dir_url)
+                                } else {
```

Comment:
The aim is to check whether the URL (in the address bar) ends with a forward-slash. If not, a forward-slash needs to be added to the end of the URL so that Servo correctly opens the links in the directory listing.

If there's a preferred way to check for the presence of an end-slash in the path, and then add one if it's missing, please point me in the right direction.

Reply:
In the latest (the fifth) revision in this pull request, I've changed this to `Url::from_directory_path(url.path())` instead, but fundamentally that's still the same thing: a URL being converted into a string and then back into a URL just to ask Servo to understand that the URL represents a directory instead of a file. If anyone knows how to refine this, let me know.